rclpy/rclpy/__init__.py

```diff
@@ -111,7 +111,9 @@ def create_node(
     namespace: str = None,
     use_global_arguments: bool = True,
     start_parameter_services: bool = True,
-    initial_parameters: List[Parameter] = None
+    initial_parameters: List[Parameter] = None,
+    allow_undeclared_parameters: bool = False,
+    automatically_declare_initial_parameters: bool = True
```

Comment:
I don't have strong opinions about it, but this option defaults to `False` in the `rclcpp` implementation. @wjwwood Thoughts?

Reply:
I know that changing this flag is technically an API change, even though it's rather small compared to the changes this PR introduces.
The only reason I see not to declare initial parameters is to define a descriptor, which would be possible if we add a `set_descriptor` method after the API freeze.
- If undeclared parameters are allowed, I don't see why not to declare initial parameters implicitly. Not doing so actually complicates the logic of `get_parameter` and `get_parameter_or` with little benefits IMO. I'd actually propose to declare them implicitly if undeclared are allowed in a follow-up.
- If undeclared parameters are not allowed and initial parameters are not declared, the user would have to re-declare all of them again. As pointed up above, the only bonus is to define a descriptor when declaring, which can be solved by introducing a `set_descriptor`.

That being said, I'd propose to go ahead and discuss what to do with the flag in the follow-up.

Reply:
I couldn't disagree more unfortunately, I think it should behave like rclcpp and default to `False`.

> If undeclared parameters are allowed, I don't see why not to declare initial parameters implicitly.

I don't see why we would declare them?

> Not doing so actually complicates the logic of get_parameter and get_parameter_or with little benefits IMO.

I disagree, if you get or set a parameter anywhere in your code, then you should declare it first. Imagine you are supposed to have a parameter which is called `foo` but you accidentally type in your code `get_parameter('fooo')`, declaring allows us to let you know that's probably not what you meant.

It just like programming in any sane language (you have to declare a variable before setting or getting it) vs CMake, where you can `${something}` and that doesn't complain even if `something` doesn't exist...

What actually complicates the logic is support `allow_undeclared_parameters`, which I only agreed to after discussing with the team and so that existing users can flip that switch and get pre-Dashing behavior.

Checking whether or not the parameter has been declared, and throwing or returning `successful=False` if it has not been declared, is easy.

> I'd actually propose to declare them implicitly if undeclared are allowed in a follow-up.

I think these two options should remain separate, and should default to off, as they only benefit the user in special use cases.

> If undeclared parameters are not allowed and initial parameters are not declared, the user would have to re-declare all of them again.

Why would they be declaring again? When did they declare it the first time? The majority of parameters, in ROS 1 anyways, are used in the node but rarely overridden externally. If you didn't declare a parameter first, then you'd always have to use `get_or` like methods to access it because you could never depend on the user to pass you all the require parameters every time. By declaring the parameter, usually in your constructor or before you call `spin`, you can access it freely through the rest of your application without worrying that it may not be set.

> As pointed up above, the only bonus is to define a descriptor when declaring, which can be solved by introducing a `set_descriptor`.

That's not the only bonus, there are also the reasons I described above, though declaring a description is a bonus to using declare parameter, especially if it should be type checked.

By not declaring them automatically, and therefore only considering the initial parameter values when you declare or set, you get to make sure the type given on the command line is valid before setting it. If you declare them automatically, then you cannot add any type checking or constraints (like range) and therefore cannot prevent invalid values from getting initialized.

Then there's the problem of setting parameters that are never used in a node. If you declare them automatically then external observers might see that you node has a parameter like `throttle` and think that it does something, when in fact your code may never even contains the string `throttle`, and in stead it was set because someone accidentally passed that parameter to your node. In a future set of changes we want to have a checker that actually throws an exception if your node is given parameters you didn't expect to help guard against typos in the initial parameter values.

So, in my opinion, there's no point in declaring parameters given by constructor or via the command line automatically, at least most of the time. Because if you don't get or set them, then they don't need to be declared in the first place, and if they're declared you don't need to set the automatically.

The only real case where I think declaring automatically is useful is something like a parameter server, where you'd have a generic executable which could store any number of parameters based on command line input and serve them to other processes, and that use case is the only reason why this option exists, from my perspective.

Maybe you have use cases in mind that we didn't consider when we discussed this in a previous ROS 2 meeting, but without that, I think we've got the right default behavior by disabling automatic declaration and requiring declaration before getting and setting.

> That being said, I'd propose to go ahead and discuss what to do with the flag in the follow-up.

We should really change it soon, because it is not a small change in behavior at all.

Reply:
Thanks for the thorough explanation!
I'll submit a PR with the change for review.

Reply:
Thanks!

Reading this again, it seems a bit aggressive, sorry about that, I've just been working on this for a while and had a lot to say on the subject. :) Hopefully, the neutral tone wasn't taken as negative.

I think the perspective you had above, is not unreasonable, and in fact we had to go in a few rounds as a team before we settled on the behavior in rclcpp.

Reply:
No worries! I've definitely found the details useful. In the rush of fixing some test cases last week I didn't have time to wrap this up as I wanted to; hopefully the changes in #339 should do the job.